warp10/src/main/java/io/warp10/script/formatted/FormattedWarpScriptFunction.java

```diff
@@ -0,0 +1,305 @@
+  //
+
+  protected abstract WarpScriptStack apply(Map<String, Object> formattedArgs, WarpScriptStack stack) throws WarpScriptException;
+
+  //
+  // Optionally, the child class can fill the fields docstring and unitTests for doc generation purpose.
+  //
+
+  protected final StringBuilder getDocstring() {
+    return docstring;
+  }
+
+  protected List<String> getUnitTests() {
+    return unitTests;
+  }
+
+  //
+  // Handle arguments parsing
+  //
+
+  /**
+   * Create the formatted args from the signatures provided by the child class, and apply child's apply method.
+   *
+   * @param stack
+   * @return stack
+   * @throws WarpScriptException
+   */
+  @Override
+  final public Object apply(WarpScriptStack stack) throws WarpScriptException {
+
+    List<ArgumentSpecification> args = getArguments().args;
+    List<ArgumentSpecification> optArgs = getArguments().optArgs;
+
+    //
+    // Sanity checks
+    //
+
+    if (null == args) {
+      throw new WarpScriptException(getClass().getCanonicalName() + "'s method getArguments() returned null. If no " +
+        "argument is expected, it should return an empty array instead.");
+    }
+
+    if (null == optArgs) {
+      throw new WarpScriptException(getClass().getCanonicalName() + "'s method getOptionalArguments() returned null." +
+        " If no argument is expected, it should return an empty array instead.");
+    }
+
+    for (ArgumentSpecification arg: args) {
+      if (arg.isOptional()) {
+        throw new IllegalStateException("Output of " + getClass().getCanonicalName() + "'s method getArguments() must" +
+          " only contain arguments without a default value.");
+      }
+    }
+
+    for (ArgumentSpecification arg: optArgs) {
+      if (!arg.isOptional()) {
+        throw new IllegalStateException("Output of " + getClass().getCanonicalName() + "'s method getArguments() must" +
+          " only contain arguments with a default value.");
+      }
+    }
+
+    //
+    // If args and opt args are empty, should expect nothing on top
+    // If args is empty but opt args is not, should expect a map (possibly empty) on top
+    //
+
+    if (0 == args.size() && 0 == optArgs.size()) {
+      return apply(new HashMap<String, Object>(), stack);
+    }
+
+    if (0 == args.size()) {
+
+      if (0 == stack.depth() || !(stack.peek() instanceof Map)) {
+        throw new WarpScriptException(getClass().getCanonicalName() + " expects a MAP on top of the stack. To use default argument values, an empty MAP is expected.");
+      }
+    }
+
+    //
+    // If there are possible optional arguments, check that last mandatory argument is not a Map so that there is no
+    // confusion
+    //
+
+    if (args.size() > 1 && args.get(args.size() - 1) instanceof Map && optArgs.size() > 0) {
+      throw new IllegalStateException("In this case, the last non-optional argument can not be a Map so as to avoid any confusion when using optional arguments.");
```

Comment:
Throw WarpScriptException

Reply:
The message starting with 'In this case' is disturbing, it looks more like a comment in the code, can you make it clearer for the user receiving it out of context.